tests/generative/variable_strategies.py

```diff
@@ -183,12 +200,84 @@ def subtract(draw, type_, frame):
     def multiply(draw, type_, frame):
         return draw(binary_operation(type_, frame, Function.Multiply))
 
+    @st.composite
+    def date_add_years(draw, type_, frame):
+        return draw(
+            binary_operation_with_types(type_, int, frame, Function.DateAddYears)
+        )
+
+    @st.composite
+    def date_add_months(draw, type_, frame):
+        return draw(
+            binary_operation_with_types(type_, int, frame, Function.DateAddMonths)
+        )
+
+    @st.composite
+    def date_add_days(draw, type_, frame):
+        return draw(
+            binary_operation_with_types(type_, int, frame, Function.DateAddDays)
+        )
+
+    @st.composite
+    def date_difference_in_years(draw, type_, frame):
+        return draw(
+            binary_operation(datetime.date, frame, Function.DateDifferenceInYears)
+        )
+
+    @st.composite
+    def date_difference_in_months(draw, type_, frame):
+        return draw(
+            binary_operation(datetime.date, frame, Function.DateDifferenceInMonths)
+        )
+
+    @st.composite
+    def date_difference_in_days(draw, type_, frame):
+        return draw(
+            binary_operation(datetime.date, frame, Function.DateDifferenceInDays)
+        )
+
     @st.composite
     def binary_operation(draw, type_, frame, operator_func):
         # A strategy for operations that take lhs and rhs arguments of the
         # same type
-        lhs = draw(series(type_, draw(one_row_per_patient_frame_or(frame))))
-        rhs = draw(series(type_, draw(one_row_per_patient_frame_or(frame))))
+        return draw(binary_operation_with_types(type_, type_, frame, operator_func))
+
+    @st.composite
+    def binary_operation_with_types(draw, lhs_type, rhs_type, frame, operator_func):
+        # A strategy for operations that take lhs and rhs arguments with specified lhs
+        # and rhs types (which may be different)
+
+        # A binary operation has 2 inputs, which are
+        # 1) A series drawn from the specified frame
+        # 2) one of:
+        #    a) A series drawn from the specified frame
+        #    b) A series drawn from a one-row-per-patient-frame
+        #    c) A Value
+
+        # Define other_frame and other_series (#2 above)
+        # pick either a one-row-per-patient-frame or this frame
+        other_frame = draw(st.one_of(one_row_per_patient_frame(), st.just(frame)))
+        if is_one_row_per_patient_frame(other_frame):
+            # if other_frame is patient-level, the input will either be a value or a
+            # series drawn from other_frame
+            other_strategy = draw(st.sampled_from([value, series]))
+        else:
+            # if other_frame is event-level, the input will be a series drawn from the
+            # other_frame
+            other_strategy = series
```

Comment:
Now I think that the comment is correct and this logic may be wrong :-). I don't think that the series strategy should vary based on whether the frame is one-row-per-patient or not, should it? There's no reason not to combine a one-row-per-patient frame with a value.